services/sky/document_view.cc

```diff
@@ -99,30 +95,64 @@ base::WeakPtr<DocumentView> DocumentView::GetWeakPtr() {
   return weak_factory_.GetWeakPtr();
 }
 
-void DocumentView::OnEmbed(
-    mojo::View* root,
-    mojo::InterfaceRequest<mojo::ServiceProvider> services_provided_to_embedder,
-    mojo::ServiceProviderPtr services_provided_by_embedder) {
-  root_ = root;
+void DocumentView::InitViewport() {
+  mojo::ServiceProviderPtr viewport_service_provider;
+  shell_->ConnectToApplication("mojo:native_viewport_service",
+                               mojo::GetProxy(&viewport_service_provider),
+                               nullptr);
+  mojo::ConnectToService(viewport_service_provider.get(), &viewport_service_);
+  viewport_service_.set_connection_error_handler(
+      [this]() { OnViewportConnectionError(); });
+
+  mojo::NativeViewportEventDispatcherPtr ptr;
+  event_dispatcher_binding_.Bind(GetProxy(&ptr));
+  viewport_service_->SetEventDispatcher(ptr.Pass());
+
+  // Match the Nexus 5 aspect ratio initially.
+  auto size = mojo::Size::New();
+  size->width = 320;
+  size->height = 640;
+
+  auto requested_configuration = mojo::SurfaceConfiguration::New();
+
+  viewport_service_->Create(size.Clone(),
+                            requested_configuration.Pass(),
+                            base::Bind(&DocumentView::OnViewportCreated,
+                                       base::Unretained(this)));
+}
 
-  if (services_provided_by_embedder.get()) {
-    mojo::ConnectToService(services_provided_by_embedder.get(),
-                           &navigator_host_);
-  }
+void DocumentView::OnViewportConnectionError() {
+  delete this;
+}
+
+void DocumentView::OnViewportCreated(mojo::ViewportMetricsPtr metrics) {
+  viewport_service_->Show();
+  mojo::ContextProviderPtr onscreen_context_provider;
+  viewport_service_->GetContextProvider(GetProxy(&onscreen_context_provider));
 
-  services_provided_to_embedder_ = services_provided_to_embedder.Pass();
-  services_provided_by_embedder_ = services_provided_by_embedder.Pass();
+  mojo::ServiceProviderPtr surfaces_service_provider;
+  shell_->ConnectToApplication("mojo:surfaces_service",
+                               mojo::GetProxy(&surfaces_service_provider),
+                               nullptr);
+  mojo::DisplayFactoryPtr display_factory;
+  mojo::ConnectToService(surfaces_service_provider.get(), &display_factory);
+  display_factory->Create(onscreen_context_provider.Pass(),
+                          nullptr, GetProxy(&display_));
 
   Load(response_.Pass());
+  UpdateViewportMetrics(*metrics);
+  RequestViewportMetrics();
+}
 
-  UpdateRootSizeAndViewportMetrics(root_->bounds());
-  root_->AddObserver(this);
+void DocumentView::OnViewportMetricsChanged(mojo::ViewportMetricsPtr metrics) {
+  UpdateViewportMetrics(*metrics);
+  RequestViewportMetrics();
```

Comment:
native_viewport promises to send an update whenever the metrics are different from those it last sent.  So I imagine we could miss intermediate updates if the metrics changed repeatedly in a short time but we should still converge on the final state which is all we care about here.